trl/trainer/ppov2_trainer.py

```diff
@@ -0,0 +1,819 @@
+        # compute_metrics: Optional[Callable[[EvalPrediction], Dict]] = None,
+        # model_init: Optional[Callable[[torch.nn.Module], None]] = None,
+        callbacks: Optional[List[TrainerCallback]] = None,
+    ) -> None:
+        self.args = args
+        self.tokenizer = tokenizer
+        self.policy = policy
+
+        self.policy.generation_config.eos_token_id = None  # disable `pad_token_id` and `eos_token_id` because we just want to
+        self.policy.generation_config.pad_token_id = None  # generate tokens without truncation / padding
+
+
+        self.ref_policy = ref_policy
+        self.reward_model = reward_model
+        self.train_dataset = train_dataset
+        self.train_dataset_len = len(train_dataset)
+        self.train_generation_config = train_generation_config
+        self.value_model = value_model
+        self.data_collator = data_collator
+        self.eval_dataset = eval_dataset
+        self.eval_generation_config = eval_generation_config
+        if eval_generation_config is None:
+            self.eval_generation_config = train_generation_config
+        self.optimizer, self.lr_scheduler = optimizers
+        self.callbacks = callbacks
+
+        #########
+        # calculate various batch sizes
+        #########
+        accelerator = Accelerator(gradient_accumulation_steps=args.gradient_accumulation_steps)
+        self.accelerator = accelerator
+        args.world_size = accelerator.num_processes
+        args.local_batch_size = args.per_device_train_batch_size * args.gradient_accumulation_steps * args.nminibatches
+        args.micro_batch_size = int(args.per_device_train_batch_size * args.world_size)
+        args.batch_size = int(args.local_batch_size * args.world_size)
+        args.mini_batch_size = exact_div(args.batch_size, args.nminibatches)
+        args.local_mini_batch_size = exact_div(args.local_batch_size, args.nminibatches)
+        if args.whiten_rewards:
+            assert (
+                args.local_mini_batch_size >= 8
+            ), f"Per-rank minibatch size {args.local_mini_batch_size} is insufficient for whitening"
+        # `per_rank_rollout_batch_size` is our `args.local_batch_size`
+        # `per_rank_minibatch_size` is our `args.local_mini_batch_size`
+        args.num_updates = args.total_episodes // args.batch_size
+        time_tensor = torch.tensor(int(time.time()), device=accelerator.device)
+        time_int = broadcast(time_tensor, 0).item()  # avoid different timestamps across processes
+        args.run_name = f"{args.exp_name}__{args.seed}__{time_int}"
+        self.local_seed = args.seed + accelerator.process_index * 100003  # Prime
+        if args.num_sample_generations > 0:
+            self.sample_generations_freq = max(1, args.num_updates // args.num_sample_generations)
+
+        #########
+        # disable dropout
+        #########
+        for module in [policy, ref_policy, value_model, reward_model]:
```

Comment:
Why are we disabling dropout for the `reward_model` and `ref_policy`?

For the trainable modules, `policy` and `value_model`, is it necessary to disable dropout? If not, I think we should leave disabling dropout to the user when they pass the modules.

Reply:
For training policy, we need dropout absolutely. Otherwise, it causes issues for PPO / REINFORCE. E.g., 


<img width="763" alt="image" src="https://github.com/huggingface/trl/assets/5555347/4c2a7cfd-7e68-4c07-8474-e1ad5770665d">


```python
import torch
import transformers
import torch.nn.functional as F
torch.set_printoptions(precision=4, sci_mode=False)


def disable_dropout(model: torch.nn.Module) -> None:
    """Disable dropout in a model."""
    for module in model.modules():
        if isinstance(module, torch.nn.Dropout):
            module.p = 0
def forward(model, query_responses, tokenizer):
    attention_mask = query_responses != tokenizer.pad_token_id
    position_ids = attention_mask.cumsum(1) - attention_mask.long()
    input_ids = torch.masked_fill(query_responses, ~attention_mask, 0)
    return model(
        input_ids=input_ids,
        attention_mask=attention_mask,
        position_ids=position_ids,
        return_dict=True,
        output_hidden_states=True,
    )

def generate(lm_backbone, queries, tokenizer, generation_config):
    """generate in a way that does not affect padding tokens"""
    context_length = queries.shape[1]
    attention_mask = queries != tokenizer.pad_token_id
    input_ids = torch.masked_fill(queries, ~attention_mask, 0)
    output = lm_backbone.generate(
        input_ids=input_ids,
        attention_mask=attention_mask,
        # position_ids=attention_mask.cumsum(1) - attention_mask.long(), # already handled in generation
        generation_config=generation_config,
        return_dict_in_generate=True,
        output_scores=True
    )
    logits = torch.stack(output.scores, 1)
    return torch.cat((queries, output.sequences[:, context_length:]), dim=1), logits


def main(tokenizer, policy, query, temperature, context_length, generation_config):
    with torch.no_grad():
        query_response, logits = generate(policy, query, tokenizer, generation_config)
        response = query_response[:, context_length:]
        all_logprob = F.log_softmax(logits, dim=-1)
        generation_logprob = torch.gather(all_logprob, 2, response.unsqueeze(-1)).squeeze(-1)
        print(f"{response=}")
        print(f"{generation_logprob=}")

    output = forward(policy, query_response, tokenizer)
    logits = output.logits[:, context_length - 1 : -1]
    logits /= temperature
    all_logprob = F.log_softmax(logits, dim=-1)
    forward_logprob = torch.gather(all_logprob, 2, response.unsqueeze(-1)).squeeze(-1)
    print(f"{forward_logprob=}")

    ppo_ratio = (forward_logprob - generation_logprob).exp()
    print(f"{ppo_ratio=}")



tokenizer = transformers.AutoTokenizer.from_pretrained("gpt2", padding_side="right")
tokenizer.add_special_tokens({"pad_token": "[PAD]"})
pad_id = tokenizer.pad_token_id
policy = transformers.AutoModelForCausalLM.from_pretrained("gpt2")
policy.generation_config.pad_token_id = policy.generation_config.eos_token_id
policy.train()
query = torch.tensor([
    [pad_id, pad_id, 23073],
    [pad_id, pad_id, 234],
])
temperature = 0.7
context_length = query.shape[1]
generation_config = transformers.GenerationConfig(
    max_new_tokens=5,
    min_new_tokens=5,
    temperature=temperature,
    top_k=0.0,
    top_p=1.0,
    do_sample=True,
)
print("=======Before disabling dropout")
main(tokenizer, policy, query, temperature, context_length, generation_config)

disable_dropout(policy)
print("=======After disabling dropout")
main(tokenizer, policy, query, temperature, context_length, generation_config)
```